src/polycap-capil.c

```diff
@@ -0,0 +1,259 @@
+	polycap_norm(surface_norm);
+
+	*alfa = acos(polycap_scalar(*surface_norm,photon_dir)); //angle between surface normal and photon direction
+	if(cos(*alfa) < 0.0){
+		return -2;
+	}
+
+	return 0;
+}
+
+//===========================================
+double polycap_refl(double e, double theta, double density, double scatf, double lin_abs_coeff){
+	// scatf = SUM( (weight/A) * (Z + f')) over all elements in capillary material
+	double complex alfa, beta; //alfa and beta component for Fresnel equation delta term (delta = alfa - i*beta)
+	double complex rtot; //reflectivity
+
+	alfa = (double)(HC/e)*(HC/e)*((N_AVOG*R0*density)/(2*M_PI)) * scatf;
+	beta = (double) (HC)/(4.*M_PI) * (lin_abs_coeff/e);
+
+	rtot = ((complex double)theta - csqrt(cpow((complex double)theta,2) - 2.*(alfa - beta*I))) / ((complex double)theta + csqrt(cpow((complex double)theta,2) - 2.*(alfa - beta*I)));
+	rtot = creal(cpow(cabs(rtot),2.));
+
+	return rtot;
+}
+
+//===========================================
+int polycap_capil_reflect(polycap_photon *photon, polycap_description *description, double alfa)
+{
+	int i, iesc=0;
+	double d_esc;  //distance in capillary at which photon escaped divided by propagation vector in z direction
+	double cons1, r_rough;
+	double complex rtot; //reflectivity
+	double w_leak; //leak weight
+	double xp, yp; //position on screen where photon will end up if unobstructed
+	int ind_x, ind_y; //indices of screen where photon will hit screen
+
+	d_esc = (description->profile->z[description->profile->nmax] - photon->exit_coords.z) / photon->exit_direction.z;
+	if(d_esc < 0) d_esc = description->profile->z[description->profile->nmax];
+	for(i=0; i < photon->n_energies; i++){
+		cons1 = (1.01358e0*photon->energies[i])*alfa*description->sig_rough;
+		r_rough = exp(-1.*cons1*cons1);
+
+		//reflectivity according to Fresnel expression
+		rtot = polycap_refl(photon->energies[i], alfa, description->density, photon->scatf[i], photon->amu[i]);
+
+		w_leak = (1.-rtot) * photon->weight[i] * exp(-1.*d_esc * photon->amu[i]);
+//		leak[i] = leak[i] + w_leak;
+		if(i == 0){ //NOTE: essentially do this for each energy to obtain photon flux image for each energy
+			xp = photon->exit_coords.x + d_esc * photon->exit_direction.x;
+			yp = photon->exit_coords.y + d_esc * photon->exit_direction.y;
+			ind_x = (int)floor(xp/BINSIZE)+NSPOT/2;
+			ind_y = (int)floor(yp/BINSIZE)+NSPOT/2;
+			if(ind_x < NSPOT && ind_x >= 0){
+				if(ind_y < NSPOT && ind_y >= 0){
+//					lspot[ind_x][ind_y] = lspot[ind_x][ind_y] + wleak;
+				}
+			}
+		}
+		photon->weight[i] = photon->weight[i] * rtot * r_rough;
+	}
+
+	if(photon->weight[0] < 1.e-4) iesc=-2;
+
+	return iesc;
+}
+
+//===========================================
+// trace photon through capillary
+int polycap_capil_trace(int *ix, polycap_photon *photon, polycap_description *description, double *cap_x, double *cap_y)
+{
+	int i, iesc=0;
+	double cap_rad0, cap_rad1;
```

Comment:
Het eerste probleem dat ik zie wordt veroorzaakt door `cap_rad1` dat nergens geinitializeerd wordt. Message in valgrind bij mij hiervoor is:

````
==17820== Conditional jump or move depends on uninitialised value(s)
==17820==    at 0x4E45930: polycap_capil_segment (polycap-capil.c:56)
==17820==    by 0x4E46848: polycap_capil_trace (polycap-capil.c:213)
==17820==    by 0x4E45431: polycap_photon_launch (polycap-photon.c:219)
==17820==    by 0x4E4440F: polycap_description_get_transmission_efficiencies (polycap-description.c:362)
==17820==    by 0x400A7E: main (main.c:36)
==17820==  Uninitialised value was created by a stack allocation
==17820==    at 0x4E4654F: polycap_capil_trace (polycap-capil.c:174)
````

_Uninitialised value was created by a stack allocation_ verwijst naar een gewone, niet ge-malloc'de variabele. Ge ziet dat het overeenkomstig lijn nummer verwijst naar het begin van die functie, dus moet je gaan kijken naar alle variabelen die in die functie gedefinieerd maar niet geinitializeerd worden.

Trek u voorlopig niet te veel aan van teveel free's: dat zouden fouten zijn die valgrind heel makkelijk kan vinden. Te weinig free's daarentegen zal waarschijnlijk hier en daar voorkomen maar dat zijn memory leaks die we later kunnen onderzoeken. In elk geval zouden die nooit de code doen crashen.

Reply:
By the way, ik kreeg hiervoor al een compiler warning:

````bash
../../src/polycap-capil.c:213:8: warning: ‘cap_rad1’ may be used uninitialized in this function [-Wmaybe-uninitialized]
   iesc = polycap_capil_segment(cap_coord0, cap_coord1, cap_rad0, cap_rad1, phot
        ^
````

Reply:
Mja, mijn compiler heeft blijkbaar een stiekem verbond gesloten om mij het leven lastig te maken, want die klaagt niet zoveel :P 

Dergelijke foutmeldingen had ik indeed ook, maar van `cap_rad0` en `rad1` snap ik het niet want die worden echt duidelijk gedefinieerd en initialiseerd (waarde gegeven) in de functie.
Kheb wel net gemerkt dat de structure `surface_norm` niet echt werd gebruikt, dus dat even aan het fiksen. Aan je bovenstaande error zou dat mss deel van het probleem kunnen zijn. Ik had idd ook de neiging om bij valgrind errors altijd naar de bovenste lijn te kijken (bij jou 56) wat me niet zo heel veel zei, maar als ik naar 174 kijk, de algemene functie, leidt het idd naar betere gevolgen... Zal het in rekening nemen in het vervolg :)

Reply:
oke, en kmoet mss een bril aanschaffen. `cap_rad1` werd inderdaad niet geinitialiseerd. Had weeral eens gewoon cap_rad0 lijn gekopieerd, en dan vergeten de 0 naar 1 veranderen :P 
Sorry voor mijn gepruts hoor...